std/ascii.zig

```diff
@@ -137,67 +138,179 @@ const combinedTable = init: {
+    inline while (i < 128) : (i += 1) {
+        table[i] =
+            u8(cntrl[i]) << @truncate(u3, @enumToInt(tIndex.Cntrl) % 8) |
+            u8(print[i]) << @truncate(u3, @enumToInt(tIndex.Print) % 8) |
+            u8(blank[i]) << @truncate(u3, @enumToInt(tIndex.Blank) % 8);
+    }
+    mem.set(u8, table[128..], 0);
+    i = 0;
+    inline while (i < 256) : (i += 1) {
+        table[i] |=
+            u8(zig[i]) << @truncate(u3, @enumToInt(tIndex.Zig) % 8);
+    }
+    break :init table;
+};
+
+inline fn inTable(c: u8, t: tIndex) bool {
+    var index = @enumToInt(t);
+    if (index <= 7) {
+        return (combinedTable[c] & (u8(1) << @truncate(u3, (index)))) != 0;
+    } else if (index <= 15) {
+        index %= 8;
+        return (combinedTable2[u8(c)] & (u8(1) << @truncate(u3, index % 8))) != 0;
+    } else unreachable;
 }
 
+/// isAlpha(c) or isDigit(c)
 pub fn isAlNum(c: u8) bool {
-    return (combinedTable[c] & ((u8(1) << @enumToInt(tIndex.Alpha)) | 
-                                 u8(1) << @enumToInt(tIndex.Digit))) != 0;
+    return (combinedTable[c] & ((u8(1) << @truncate(u3, @enumToInt(tIndex.Alpha))) | 
+                                 u8(1) << @truncate(u3, @enumToInt(tIndex.Digit)))) != 0;
 }
 
+/// 'a'...'z', 'A'...'Z'
 pub fn isAlpha(c: u8) bool {
     return inTable(c, tIndex.Alpha);
 }
 
+/// c < 0x20 or c == 127 (DEL)
 pub fn isCntrl(c: u8) bool {
-    return c < 0x20 or c == 127; //DEL
+    return inTable(c, tIndex.Cntrl);
 }
 
+/// '0'...'9'
 pub fn isDigit(c: u8) bool {
     return inTable(c, tIndex.Digit);
 }
 
+/// !isCtrl(c) and c != 0x20 (space)
 pub fn isGraph(c: u8) bool {
     return inTable(c, tIndex.Graph);
 }
 
+/// 'a'...'z'
 pub fn isLower(c: u8) bool {
     return inTable(c, tIndex.Lower);
 }
 
+/// !isCtrl(c)
 pub fn isPrint(c: u8) bool {
-    return inTable(c, tIndex.Graph) or c == ' ';
+    return inTable(c, tIndex.Print);
 }
 
+/// !isGraph(c) and !isAlNum(c)
 pub fn isPunct(c: u8) bool {
     return inTable(c, tIndex.Punct);
 }
 
+/// space, form-feed, newline, carriage return, horizontal tab, and vertical tab
 pub fn isSpace(c: u8) bool {
     return inTable(c, tIndex.Space);
 }
 
+/// 'A'...'Z'
 pub fn isUpper(c: u8) bool {
     return inTable(c, tIndex.Upper);
 }
 
+/// '0'...'9' or 'a'...'f' or 'A'...'F'
 pub fn isXDigit(c: u8) bool {
     return inTable(c, tIndex.Hex);
 }
 
+/// c < 128
 pub fn isASCII(c: u8) bool {
     return c < 128;
 }
 
+/// space or horizontal tab
 pub fn isBlank(c: u8) bool {
-    return (c == ' ') or (c == '\x09');
+    return inTable(c, tIndex.Blank);
+}
+
+/// Is a byte forbidden in Zig source code.
+/// Specifically, invalid UTF-8 bytes, and ASCII
+/// control characters except newline.
+/// see doc/langref.html.in online at https://ziglang.org/documentation/master/#Source-Encoding
+/// Does not validate UTF-8 or check for prohibited Unicode code-points,
+/// is why it is called isntZig() rather than isZig().
+pub fn isntZig(c: u8) bool {
```

Comment:
What's the reasoning behind adding this? It is surprising that there would be `std.ascii.isntZig`. What is the intended usage? Self-hosted tokenizer? What would a different language tokenizer be expected to do, since there wouldn't be, for example, std.ascii.java, std.ascii.perl, etc.?

Reply:
> Self-hosted tokenizer?

Yes.

Reply:
Putting it here saves 256 bytes. It isn't much, but it is something. If it is too ugly, then so be it. Zig is unlikely to ever tokenize java or perl, and no other language I know of has character requirements quite like zig. for C you need to support tri-graphs for example.

Reply:
Also, by putting this stuff together, a future vectored streaming version can all use the same code.

Reply:
i don't see a reason to expose this in the stdlib. if it's for parsing zig it should go in std.zig.parse.